From 65d7fc3c63e4041b82d080a562ac95dac651621f Mon Sep 17 00:00:00 2001
From: Herbert Xu <herbert.xu@redhat.com>
Date: Thu, 11 Dec 2014 13:07:57 -0500
Subject: [net] ipsec: Don't update the pmtu on ICMPV6_DEST_UNREACH

Message-id: <20141211130757.GA12796@gondor.apana.org.au>
Patchwork-id: 101655
O-Subject: [RHEL7 PATCH] ipsec: Don't update the pmtu on ICMPV6_DEST_UNREACH
Bugzilla: 1158771
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Flavio Leitner <fbl@redhat.com>
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

RHEL7 Bugzilla #1158771

This patch fixes a RHEL7 regression that breaks USGv6 testing.

commit b3b2b9e192d5811f91f9cd92aeec489cecabc92e
Author: Steffen Klassert <steffen.klassert@secunet.com>
Date:   Tue Sep 10 13:43:09 2013 +0200

    ipsec: Don't update the pmtu on ICMPV6_DEST_UNREACH

    Currently we update the pmtu in the IPsec protocol error handlers
    if icmpv6 message type is either ICMPV6_DEST_UNREACH or
    ICMPV6_PKT_TOOBIG. Updating the pmtu on ICMPV6_DEST_UNREACH
    is wrong in any case, it causes strangely fragmented packets.
    Only ICMPV6_PKT_TOOBIG signalizes pmtu discovery, so remove the
    ICMPV6_DEST_UNREACH check in the IPsec protocol error handlers.

    Signed-off-by: Steffen Klassert <steffen.klassert@secunet.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv6/ah6.c b/net/ipv6/ah6.c
index 73784c3..82e1da3 100644
--- a/net/ipv6/ah6.c
+++ b/net/ipv6/ah6.c
@@ -618,8 +618,7 @@ static void ah6_err(struct sk_buff *skb, struct inet6_skb_parm *opt,
  struct ip_auth_hdr *ah = (struct ip_auth_hdr*)(skb->data+offset);
  struct xfrm_state *x;
 
- if (type != ICMPV6_DEST_UNREACH &&
-     type != ICMPV6_PKT_TOOBIG &&
+ if (type != ICMPV6_PKT_TOOBIG &&
      type != NDISC_REDIRECT)
   return;
 
diff --git a/net/ipv6/esp6.c b/net/ipv6/esp6.c
index b8b5a15..a2c9d0e 100644
--- a/net/ipv6/esp6.c
+++ b/net/ipv6/esp6.c
@@ -436,8 +436,7 @@ static void esp6_err(struct sk_buff *skb, struct inet6_skb_parm *opt,
  struct ip_esp_hdr *esph = (struct ip_esp_hdr *)(skb->data + offset);
  struct xfrm_state *x;
 
- if (type != ICMPV6_DEST_UNREACH &&
-     type != ICMPV6_PKT_TOOBIG &&
+ if (type != ICMPV6_PKT_TOOBIG &&
      type != NDISC_REDIRECT)
   return;
 
diff --git a/net/ipv6/ipcomp6.c b/net/ipv6/ipcomp6.c
index 5636a91..ce507d9 100644
--- a/net/ipv6/ipcomp6.c
+++ b/net/ipv6/ipcomp6.c
@@ -64,8 +64,7 @@ static void ipcomp6_err(struct sk_buff *skb, struct inet6_skb_parm *opt,
   (struct ip_comp_hdr *)(skb->data + offset);
  struct xfrm_state *x;
 
- if (type != ICMPV6_DEST_UNREACH &&
-     type != ICMPV6_PKT_TOOBIG &&
+ if (type != ICMPV6_PKT_TOOBIG &&
      type != NDISC_REDIRECT)
   return;
 
-- 
1.7.1